geomstats/discretized_curves_space.py

```diff
@@ -0,0 +1,437 @@
+                                base_curve):
+        """
+        Compute the inner products of the components of a (series of)
+        pair(s) of tangent vectors at (a) base curve(s).
+        """
+        base_curve = gs.to_ndarray(base_curve, to_ndim=3)
+        tangent_vec_a = gs.to_ndarray(tangent_vec_a, to_ndim=3)
+        tangent_vec_b = gs.to_ndarray(tangent_vec_b, to_ndim=3)
+
+        n_tangent_vecs = tangent_vec_a.shape[0]
+        n_sampling_points = tangent_vec_a.shape[1]
+        inner_prod = gs.zeros([n_tangent_vecs, n_sampling_points])
+        for k in range(n_tangent_vecs):
+            inner_prod[k, :] = self.embedding_metric.inner_product(
+                    tangent_vec_a[k, :], tangent_vec_b[k, :],
+                    base_curve[k, :]).squeeze()
+
+        return inner_prod
+
+    def pointwise_norm(self, tangent_vec, base_curve):
+        """
+        Compute the norms of the components of a (series of)
+        tangent vector(s) at (a) base curve(s).
+        """
+        sq_norm = self.pointwise_inner_product(
+                tangent_vec_a=tangent_vec,
+                tangent_vec_b=tangent_vec,
+                base_curve=base_curve)
+        return gs.sqrt(sq_norm)
+
+    def square_root_velocity(self, curve):
+        """
+        Compute the square root velocity representation of a curve.
+        """
+        srv_shape = gs.array(curve.shape)
+        srv_shape[-2] -= 1
+
+        curve = gs.to_ndarray(curve, to_ndim=3)
+        n_curves = curve.shape[0]
+        n_sampling_points = curve.shape[1]
+        n_coords = curve.shape[2]
+
+        curve = curve.reshape(n_curves * n_sampling_points, n_coords)
+        velocity = n_sampling_points * self.embedding_metric.log(
+                point=curve[1:, :], base_point=curve[:-1, :])
+        velocity_norm = self.embedding_metric.norm(velocity, curve[:-1, :])
+        assert gs.all(velocity_norm != 0)
+        srv = velocity / gs.sqrt(velocity_norm)
+
+        index = gs.arange(n_curves * n_sampling_points - 1)
+        index_select = index[(index + 1) % n_sampling_points != 0]
+        srv = srv[index_select, :].reshape(srv_shape)
+
+        return srv
+
+    def square_root_velocity_inverse(self, srv, starting_point):
+        """
+        Retreive a curve from its square root velocity representation
+        and starting point.
+        """
+        if not isinstance(self.embedding_metric, EuclideanMetric):
+            raise AssertionError('The square root velocity inverse is only '
+                                 'implemented for dicretized curves embedded '
+                                 'in a Euclidean space.')
+        if srv.ndim != starting_point.ndim:
+            starting_point = gs.transpose(
+                    np.tile(starting_point, (1, 1, 1)), (1, 0, 2)
+                    )
+        srv_shape = srv.shape
+        srv = gs.to_ndarray(srv, to_ndim=3)
+        n_curves = srv.shape[0]
+        n_sampling_points = srv.shape[1] + 1
```

Comment:
why +1?

Reply:
To be consistent with the names of the other methods : ```n_sampling_points``` is the number of sampling points of the curve, and so when we deal with its derivative or srv (computed with finite differences) we 'loose' a point. I suggest to rename it ```n_sampling_points_minus_one``` ?

This made me realize I should be multiplying by ```n_sampling_points - 1``` rather than ```n_sampling_points``` everywhere where I compute velocities (the curves are supposed to be parameterized between 0 and 1, at regular intervals of 1/(n-1) if there are n points).